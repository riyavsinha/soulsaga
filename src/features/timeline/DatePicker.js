import FormControl from '@material-ui/core/FormControl';
import FormHelperText from '@material-ui/core/FormHelperText';
import MenuItem from '@material-ui/core/MenuItem';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import TextField from '@material-ui/core/TextField';

const MONTHS = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December',
];

/**
 * Fully Controlled Component - AddEventForm
 */
export default class DatePicker extends Component {
  static propTypes = {
    day: PropTypes.string.isRequired,
    month: PropTypes.oneOf(MONTHS.concat('')).isRequired,
    year: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
  };

  /**
   * Determines whether a day/month/year date is invalid.
   *
   * Valid dates are either just a year, a year with a month, or a year with
   * month and day.
   *
   * An invalid date is therefore determined in one of 3 ways:
   *  (1) No year
   *  (2) A year with a day but no month
   *  (3) A year with a month and day, but day is invalid for given month/year
   *
   * @param y {!String} The date's year
   * @param m {!String} The date's month
   * @param d {!String} The date's day
   * @param includeYearError {bool=} Whether to calculate invalid date
   *  including invalid case (1). Used in rendered component since
   *  case (1) has separate error from cases (2) and (3).
   */
  isInvalidDate = (y, m, d, includeYearError = true) => {
    const numericReg = new RegExp('^\\d*$');
    const noYear = !y;
    const nonNumericYear = !numericReg.test(y);
    const dayWithoutMonth = !!(y && d && !m);
    const dayOutOfBounds = !!(
      y &&
      d &&
      m &&
      (d > new Date(y, MONTHS.indexOf(m) + 1, 0).getDate() || d < 1)
    );
    const nonNumericDay = !numericReg.test(d);
    const invalidWithoutYearError =
      nonNumericYear || dayWithoutMonth || dayOutOfBounds || nonNumericDay;
    return includeYearError
      ? noYear || invalidWithoutYearError
      : invalidWithoutYearError;
  };

  blockNonNumericKeys = e => {
    if (['+', '.', '-', 'e'].includes(e.key)) {
      e.preventDefault();
    }
  };

  handleDateChange = (d, m, y) =>
    this.props.onChange(m, d, y, !this.isInvalidDate(y, m, d, true));

  handleMonthChange = e =>
    this.handleDateChange(this.props.day, e.target.value, this.props.year);

  handleDayChange = e =>
    this.handleDateChange(e.target.value, this.props.month, this.props.year);

  handleYearChange = e =>
    this.handleDateChange(this.props.day, this.props.month, e.target.value);

  render() {
    // Shorthand props
    const d = this.props.day;
    const m = this.props.month;
    const y = this.props.year;
    // Determine errors
    const hasNoYearError = !y;
    const isInvalidDateWithYear =
      this.isInvalidDate(y, m, d, false) && !hasNoYearError;

    return (
      <FormControl>
        <div>
          {/** Month Select */}
          <TextField
            select
            label="Month"
            value={m}
            onChange={this.handleMonthChange}
            className="timeline-date-picker__month timeline-date-picker__field"
            margin="normal"
            SelectProps={{
              MenuProps: {
                className: 'timeline-date-picker__month-select-menu',
              },
            }}
          >
            <MenuItem key="none" value="" />
            {MONTHS.map(month => (
              <MenuItem key={month} value={month}>
                {month}
              </MenuItem>
            ))}
          </TextField>

          {/** Day Number Input */}
          <TextField
            label="Day"
            value={d}
            type="number"
            onChange={this.handleDayChange}
            inputProps={{ onKeyDown: this.blockNonNumericKeys }}
            className="timeline-date-picker__field"
            InputLabelProps={{
              shrink: true,
            }}
            margin="normal"
          />

          {/** Year Number Input */}
          <TextField
            required
            label="Year"
            value={y}
            type="number"
            onChange={this.handleYearChange}
            inputProps={{ onKeyDown: this.blockNonNumericKeys }}
            InputLabelProps={{
              shrink: true,
            }}
            margin="normal"
            error={hasNoYearError}
            helperText={hasNoYearError ? 'This field is required' : ''}
          />
        </div>

        <FormHelperText
          error={isInvalidDateWithYear}
          className={
            isInvalidDateWithYear
              ? ''
              : 'timeline-date-picker__date-error--hidden'
          }
        >
          Please enter a valid date
        </FormHelperText>
      </FormControl>
    );
  }
}
