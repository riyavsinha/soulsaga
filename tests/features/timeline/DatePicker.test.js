import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import DatePicker from 'features/timeline/DatePicker';
import Input from '@material-ui/core/Input';

const NO_YEAR_EROR = 'This field is required';
const INVALID_DATE_HIDDEN_CLASS = 'timeline-date-picker__date-error--hidden';

describe('timeline/DatePicker', () => {
  let onChangeFn;

  beforeEach(() => {
    onChangeFn = jest.fn();
  });

  describe('renders', () => {
    it('renders empty', () => {
      const props = { month: '', day: '', year: '', onChange: onChangeFn };
      const renderedComponent = renderer
        .create(<DatePicker {...props} />)
        .toJSON();

      expect(renderedComponent).toMatchSnapshot();
    });

    it('renders with date', () => {
      const props = {
        month: 'January',
        day: '5',
        year: '2010',
        onChange: onChangeFn,
      };
      const renderedComponent = renderer
        .create(<DatePicker {...props} />)
        .toJSON();

      expect(renderedComponent).toMatchSnapshot();
    });

    it('renders with invalid date', () => {
      const props = { month: '', day: '5', year: '2010', onChange: onChangeFn };
      const renderedComponent = renderer
        .create(<DatePicker {...props} />)
        .toJSON();

      expect(renderedComponent).toMatchSnapshot();
    });
  });

  describe('"Year Required" error is shown when no year', () => {
    let props;
    beforeEach(() => {
      props = {
        month: '',
        day: '',
        year: '',
        onChange: onChangeFn,
      };
    });

    it('has empty props', () => {
      const wrapper = mount(<DatePicker {...props} />);

      expect(
        wrapper
          .find('p')
          .first()
          .text(),
      ).toBe(NO_YEAR_EROR);
      expect(
        wrapper
          .find('p')
          .last()
          .hasClass(INVALID_DATE_HIDDEN_CLASS),
      ).toBe(true);
    });

    it('has month', () => {
      props.month = 'January';
      const wrapper = mount(<DatePicker {...props} />);

      expect(
        wrapper
          .find('p')
          .first()
          .text(),
      ).toBe(NO_YEAR_EROR);
      expect(
        wrapper
          .find('p')
          .last()
          .hasClass(INVALID_DATE_HIDDEN_CLASS),
      ).toBe(true);
    });

    it('has day', () => {
      props.day = '5';
      const wrapper = mount(<DatePicker {...props} />);

      expect(
        wrapper
          .find('p')
          .first()
          .text(),
      ).toBe(NO_YEAR_EROR);
      expect(
        wrapper
          .find('p')
          .last()
          .hasClass(INVALID_DATE_HIDDEN_CLASS),
      ).toBe(true);
    });

    it('has month and day', () => {
      props.month = 'January';
      props.day = '5';
      const wrapper = mount(<DatePicker {...props} />);

      expect(
        wrapper
          .find('p')
          .first()
          .text(),
      ).toBe(NO_YEAR_EROR);
      expect(
        wrapper
          .find('p')
          .last()
          .hasClass(INVALID_DATE_HIDDEN_CLASS),
      ).toBe(true);
    });

    it('hides invalid date error when no year', () => {
      props.month = 'March';
      props.day = '40';
      const wrapper = mount(<DatePicker {...props} />);

      expect(
        wrapper
          .find('p')
          .last()
          .hasClass(INVALID_DATE_HIDDEN_CLASS),
      ).toBe(true);
    });
  });

  describe('"Invalid Date" error is shown when bad date', () => {
    let props;
    beforeEach(() => {
      props = {
        month: '',
        day: '',
        year: '2001',
        onChange: onChangeFn,
      };
    });

    it('has day with no month', () => {
      props.day = '4';
      const wrapper = mount(<DatePicker {...props} />);

      expect(wrapper.find('p')).toHaveLength(1);
      expect(
        wrapper
          .find('p')
          .first()
          .hasClass(INVALID_DATE_HIDDEN_CLASS),
      ).toBe(false);
    });

    it('day is too low', () => {
      props.day = '0';
      const wrapper = mount(<DatePicker {...props} />);

      expect(wrapper.find('p')).toHaveLength(1);
      expect(
        wrapper
          .find('p')
          .first()
          .hasClass(INVALID_DATE_HIDDEN_CLASS),
      ).toBe(false);
    });

    it('day is too high', () => {
      props.day = '29';
      props.month = 'February';
      const wrapper = mount(<DatePicker {...props} />);

      expect(wrapper.find('p')).toHaveLength(1);
      expect(
        wrapper
          .find('p')
          .first()
          .hasClass(INVALID_DATE_HIDDEN_CLASS),
      ).toBe(false);
    });
  });

  describe('Valid date shows no errors', () => {
    let props;
    beforeEach(() => {
      props = {
        month: '',
        day: '',
        year: '2000',
        onChange: onChangeFn,
      };
    });

    it('has year', () => {
      const wrapper = mount(<DatePicker {...props} />);

      expect(wrapper.find('p')).toHaveLength(1);
      expect(
        wrapper
          .find('p')
          .first()
          .hasClass(INVALID_DATE_HIDDEN_CLASS),
      ).toBe(true);
    });

    it('has year and month', () => {
      props.month = 'July';
      const wrapper = mount(<DatePicker {...props} />);

      expect(wrapper.find('p')).toHaveLength(1);
      expect(
        wrapper
          .find('p')
          .first()
          .hasClass(INVALID_DATE_HIDDEN_CLASS),
      ).toBe(true);
    });

    it('has year, month and day', () => {
      props.month = 'July';
      props.day = '31';
      const wrapper = mount(<DatePicker {...props} />);

      expect(wrapper.find('p')).toHaveLength(1);
      expect(
        wrapper
          .find('p')
          .first()
          .hasClass(INVALID_DATE_HIDDEN_CLASS),
      ).toBe(true);
    });

    it('works on leap days', () => {
      props.month = 'February';
      props.day = '29';
      const wrapper = mount(<DatePicker {...props} />);

      expect(wrapper.find('p')).toHaveLength(1);
      expect(
        wrapper
          .find('p')
          .first()
          .hasClass(INVALID_DATE_HIDDEN_CLASS),
      ).toBe(true);
    });
  });

  describe('Handles events properly', () => {
    it('updates day on change', () => {
      const props = {
        month: 'January',
        day: '5',
        year: '2010',
        onChange: onChangeFn,
      };
      const wrapper = mount(<DatePicker {...props} />);

      wrapper
        .find('input')
        .at(1)
        .simulate('change', { target: { value: '1' } });

      expect(onChangeFn.mock.calls.length).toBe(1);
      expect(onChangeFn.mock.calls[0]).toMatchObject([
        'January',
        '1',
        '2010',
        true,
      ]);
    });

    it('updates year on change', () => {
      const props = {
        month: 'January',
        day: '5',
        year: '2010',
        onChange: onChangeFn,
      };
      const wrapper = mount(<DatePicker {...props} />);

      wrapper
        .find('input')
        .at(2)
        .simulate('change', { target: { value: '2012' } });

      expect(onChangeFn.mock.calls.length).toBe(1);
      expect(onChangeFn.mock.calls[0]).toMatchObject([
        'January',
        '5',
        '2012',
        true,
      ]);
    });

    it('updates month on change', () => {
      const props = {
        month: 'January',
        day: '5',
        year: '2010',
        onChange: onChangeFn,
      };
      const wrapper = mount(<DatePicker {...props} />);

      // Open menu
      wrapper.find('[role="button"]').simulate('click');
      // Click menu option
      wrapper
        .find('[data-value="August"]')
        .first()
        .simulate('click');

      expect(onChangeFn.mock.calls.length).toBe(1);
      expect(onChangeFn.mock.calls[0]).toMatchObject([
        'August',
        '5',
        '2010',
        true,
      ]);
    });
  });
});
