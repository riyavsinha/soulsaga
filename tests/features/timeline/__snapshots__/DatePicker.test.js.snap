// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`timeline/DatePicker renders renders empty 1`] = `
<div
  className="MuiFormControl-root-1"
>
  <div>
    <div
      className="MuiFormControl-root-1 MuiFormControl-marginNormal-2 timeline-date-picker__month timeline-date-picker__field"
    >
      <label
        className="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13"
        data-shrink={false}
      >
        Month
      </label>
      <div
        className="MuiInputBase-root-43 MuiInput-root-30 MuiInput-underline-34 MuiInputBase-formControl-44 MuiInput-formControl-31"
        onClick={[Function]}
      >
        <div
          className="MuiSelect-root-23"
        >
          <div
            aria-haspopup="true"
            aria-pressed="false"
            className="MuiSelect-select-24 MuiSelect-selectMenu-27 MuiInputBase-input-53 MuiInput-input-38"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={0}
          >
            <span
              dangerouslySetInnerHTML={
                Object {
                  "__html": "&#8203;",
                }
              }
            />
          </div>
          <input
            type="hidden"
            value=""
          />
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-60 MuiSelect-icon-29"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      className="MuiFormControl-root-1 MuiFormControl-marginNormal-2 timeline-date-picker__field"
    >
      <label
        className="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-shrink-12"
        data-shrink={true}
      >
        Day
      </label>
      <div
        className="MuiInputBase-root-43 MuiInput-root-30 MuiInput-underline-34 MuiInputBase-formControl-44 MuiInput-formControl-31"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          className="MuiInputBase-input-53 MuiInput-input-38 MuiInputBase-inputType-56 MuiInput-inputType-41"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          required={false}
          type="number"
          value=""
        />
      </div>
    </div>
    <div
      className="MuiFormControl-root-1 MuiFormControl-marginNormal-2"
    >
      <label
        className="MuiFormLabel-root-16 MuiFormLabel-error-19 MuiInputLabel-error-8 MuiFormLabel-required-21 MuiInputLabel-required-9 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-shrink-12"
        data-shrink={true}
      >
        Year
        <span
          className="MuiFormLabel-asterisk-22 MuiFormLabel-error-19 MuiInputLabel-error-8"
        >
           *
        </span>
      </label>
      <div
        className="MuiInputBase-root-43 MuiInput-root-30 MuiInput-underline-34 MuiInputBase-error-49 MuiInput-error-35 MuiInputBase-formControl-44 MuiInput-formControl-31"
        onClick={[Function]}
      >
        <input
          aria-invalid={true}
          className="MuiInputBase-input-53 MuiInput-input-38 MuiInputBase-inputType-56 MuiInput-inputType-41"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          required={true}
          type="number"
          value=""
        />
      </div>
      <p
        className="MuiFormHelperText-root-73 MuiFormHelperText-error-74 MuiFormHelperText-required-80"
      >
        This field is required
      </p>
    </div>
  </div>
  <p
    className="MuiFormHelperText-root-73 timeline-date-picker__date-error--hidden"
  >
    Please enter a valid date
  </p>
</div>
`;

exports[`timeline/DatePicker renders renders with date 1`] = `
<div
  className="MuiFormControl-root-1"
>
  <div>
    <div
      className="MuiFormControl-root-1 MuiFormControl-marginNormal-2 timeline-date-picker__month timeline-date-picker__field"
    >
      <label
        className="MuiFormLabel-root-16 MuiFormLabel-filled-20 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-shrink-12"
        data-shrink={true}
      >
        Month
      </label>
      <div
        className="MuiInputBase-root-43 MuiInput-root-30 MuiInput-underline-34 MuiInputBase-formControl-44 MuiInput-formControl-31"
        onClick={[Function]}
      >
        <div
          className="MuiSelect-root-23"
        >
          <div
            aria-haspopup="true"
            aria-pressed="false"
            className="MuiSelect-select-24 MuiSelect-selectMenu-27 MuiInputBase-input-53 MuiInput-input-38"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={0}
          >
            January
          </div>
          <input
            type="hidden"
            value="January"
          />
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-60 MuiSelect-icon-29"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      className="MuiFormControl-root-1 MuiFormControl-marginNormal-2 timeline-date-picker__field"
    >
      <label
        className="MuiFormLabel-root-16 MuiFormLabel-filled-20 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-shrink-12"
        data-shrink={true}
      >
        Day
      </label>
      <div
        className="MuiInputBase-root-43 MuiInput-root-30 MuiInput-underline-34 MuiInputBase-formControl-44 MuiInput-formControl-31"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          className="MuiInputBase-input-53 MuiInput-input-38 MuiInputBase-inputType-56 MuiInput-inputType-41"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          required={false}
          type="number"
          value="5"
        />
      </div>
    </div>
    <div
      className="MuiFormControl-root-1 MuiFormControl-marginNormal-2"
    >
      <label
        className="MuiFormLabel-root-16 MuiFormLabel-filled-20 MuiFormLabel-required-21 MuiInputLabel-required-9 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-shrink-12"
        data-shrink={true}
      >
        Year
        <span
          className="MuiFormLabel-asterisk-22"
        >
           *
        </span>
      </label>
      <div
        className="MuiInputBase-root-43 MuiInput-root-30 MuiInput-underline-34 MuiInputBase-formControl-44 MuiInput-formControl-31"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          className="MuiInputBase-input-53 MuiInput-input-38 MuiInputBase-inputType-56 MuiInput-inputType-41"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          required={true}
          type="number"
          value="2010"
        />
      </div>
      
    </div>
  </div>
  <p
    className="MuiFormHelperText-root-73 timeline-date-picker__date-error--hidden"
  >
    Please enter a valid date
  </p>
</div>
`;

exports[`timeline/DatePicker renders renders with invalid date 1`] = `
<div
  className="MuiFormControl-root-1"
>
  <div>
    <div
      className="MuiFormControl-root-1 MuiFormControl-marginNormal-2 timeline-date-picker__month timeline-date-picker__field"
    >
      <label
        className="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13"
        data-shrink={false}
      >
        Month
      </label>
      <div
        className="MuiInputBase-root-43 MuiInput-root-30 MuiInput-underline-34 MuiInputBase-formControl-44 MuiInput-formControl-31"
        onClick={[Function]}
      >
        <div
          className="MuiSelect-root-23"
        >
          <div
            aria-haspopup="true"
            aria-pressed="false"
            className="MuiSelect-select-24 MuiSelect-selectMenu-27 MuiInputBase-input-53 MuiInput-input-38"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={0}
          >
            <span
              dangerouslySetInnerHTML={
                Object {
                  "__html": "&#8203;",
                }
              }
            />
          </div>
          <input
            type="hidden"
            value=""
          />
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-60 MuiSelect-icon-29"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      className="MuiFormControl-root-1 MuiFormControl-marginNormal-2 timeline-date-picker__field"
    >
      <label
        className="MuiFormLabel-root-16 MuiFormLabel-filled-20 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-shrink-12"
        data-shrink={true}
      >
        Day
      </label>
      <div
        className="MuiInputBase-root-43 MuiInput-root-30 MuiInput-underline-34 MuiInputBase-formControl-44 MuiInput-formControl-31"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          className="MuiInputBase-input-53 MuiInput-input-38 MuiInputBase-inputType-56 MuiInput-inputType-41"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          required={false}
          type="number"
          value="5"
        />
      </div>
    </div>
    <div
      className="MuiFormControl-root-1 MuiFormControl-marginNormal-2"
    >
      <label
        className="MuiFormLabel-root-16 MuiFormLabel-filled-20 MuiFormLabel-required-21 MuiInputLabel-required-9 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-shrink-12"
        data-shrink={true}
      >
        Year
        <span
          className="MuiFormLabel-asterisk-22"
        >
           *
        </span>
      </label>
      <div
        className="MuiInputBase-root-43 MuiInput-root-30 MuiInput-underline-34 MuiInputBase-formControl-44 MuiInput-formControl-31"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          className="MuiInputBase-input-53 MuiInput-input-38 MuiInputBase-inputType-56 MuiInput-inputType-41"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          required={true}
          type="number"
          value="2010"
        />
      </div>
      
    </div>
  </div>
  <p
    className="MuiFormHelperText-root-73 MuiFormHelperText-error-74"
  >
    Please enter a valid date
  </p>
</div>
`;
